src/hotspot/share/oops/klass.cpp

```diff
@@ -163,24 +163,46 @@ void Klass::release_C_heap_structures(bool release_constant_pool) {
   if (_name != nullptr) _name->decrement_refcount();
 }
 
-bool Klass::search_secondary_supers(Klass* k) const {
-  // Put some extra logic here out-of-line, before the search proper.
-  // This cuts down the size of the inline method.
-
-  // This is necessary, since I am never in my own secondary_super list.
-  if (this == k)
-    return true;
+bool Klass::linear_search_secondary_supers(const Klass* k) const {
   // Scan the array-of-objects for a match
+  // FIXME: We could do something smarter here, maybe a vectorized
+  // comparison or a binary search, but is that worth any added
+  // complexity?
   int cnt = secondary_supers()->length();
   for (int i = 0; i < cnt; i++) {
     if (secondary_supers()->at(i) == k) {
-      ((Klass*)this)->set_secondary_super_cache(k);
       return true;
     }
   }
   return false;
 }
 
+// Given a secondary superklass k, an initial array index, and an
+// occupancy bitmap rotated such that Bit 1 is the next bit to test,
+// search for k.
+bool Klass::fallback_search_secondary_supers(const Klass* k, int index, uintx rotated_bitmap) const {
+  if (secondary_supers()->length() > SECONDARY_SUPERS_TABLE_SIZE - 2) {
+    return linear_search_secondary_supers(k);
+  }
+
+  // This is conventional linear probing, but instead of terminating
+  // when a null entry is found in the table, we maintain a bitmap
+  // in which a 0 indicates missing entries.
+
+  // The check at the start of this function guarantees there are 0s
+  // in the bitmap, so this loop eventually terminates.
```

Comment:
I don't see the check that this comment is referring to.

Reply:
I guess it means the check at line 184, which on the surface doesn't involve the bitmap at all, but I guess there is a directly relationship between secondary_supers()->length() and the bitmap.  It wouldn't hurt to spell this out more clearly.